File: extensions/stable-api-sandbox/ZooProcessor.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ZooProcessor.h"

#include "AnimalControllerServices.h"
#include "api/core/ProcessContext.h"

namespace org::apache::nifi::minifi::api_sandbox {

MinifiStatus ZooProcessor::onTriggerImpl(api::core::ProcessContext& process_context, api::core::ProcessSession& process_session) {
  if (const auto can_fly_opaque = process_context.getControllerService(CanFlyService)) {
    if (*can_fly_opaque) {
      const auto can_fly_controller_name = process_context.getProperty(CanFlyService, nullptr) | utils::orThrow("Should be here");
      const CanFlyControllerApi* can_fly = reinterpret_cast<CanFlyControllerApi*>(*can_fly_opaque);
      logger_->log_critical("Can {} fly? {}", can_fly_controller_name, can_fly->canFly());
    }
  }
  if (const auto num_of_legs_opaque = process_context.getControllerService(NumberOfLegsService)) {
    if (*num_of_legs_opaque) {
      const auto num_of_legs_name = process_context.getProperty(NumberOfLegsService, nullptr) | utils::orThrow("Should be here");
      const NumberOfLegsControllerApi* num_legs = reinterpret_cast<NumberOfLegsControllerApi*>(*num_of_legs_opaque);
      logger_->log_critical("{} has {} legs", num_of_legs_name, num_legs->numberOfLegs());
    }
  }
  return ProcessorImpl::onTriggerImpl(process_context, process_session);
}

MinifiStatus ZooProcessor::onScheduleImpl(api::core::ProcessContext& process_context) {
  return ProcessorImpl::onScheduleImpl(process_context);
}
}  // namespace org::apache::nifi::minifi::api_sandbox

File: minifi_rust/extensions/minifi_rs_playground/src/processors/zoo_processor.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::controller_services::animal_controller_apis::{
    CanFlyControllerApi, NumberOfLegsControllerApi,
};
use minifi_native::macros::ComponentIdentifier;
use minifi_native::{
    GetProperty, Logger, MinifiError, OnTriggerResult, OutputAttribute, ProcessContext,
    ProcessError, ProcessSession, ProcessorDefinition, ProcessorInputRequirement, Property,
    PropertyDefinition, Relationship, Schedule, Trigger, critical, info, property_definitions,
};

pub(crate) const CAN_FLY_SERVICE: Property<dyn CanFlyControllerApi> =
    Property::new("Can fly service", "Test CanFlyService");

pub(crate) const NUMBER_OF_LEGS: Property<dyn NumberOfLegsControllerApi> =
    Property::new("Number of Legs service", "Test NumberOfLegsService");

#[derive(Debug, ComponentIdentifier)]
pub(crate) struct ZooProcessorRs {}

impl Schedule for ZooProcessorRs {
    fn schedule<Ctx: GetProperty, L: Logger>(
        _context: &Ctx,
        _logger: &L,
    ) -> Result<Self, MinifiError>
    where
        Self: Sized,
    {
        Ok(Self {})
    }
}

impl Trigger for ZooProcessorRs {
    fn trigger<Context, Session, Lggr>(
        &self,
        context: &mut Context,
        _session: &mut Session,
        logger: &Lggr,
    ) -> Result<OnTriggerResult, ProcessError>
    where
        Context: ProcessContext,
        Session: ProcessSession<FlowFile = Context::FlowFile>,
        Lggr: Logger,
    {
        info!(logger, "{:?}", self);
        if let Some(maybe_flyer) = context.get_controller_service_api(&CAN_FLY_SERVICE)? {
            critical!(
                logger,
                "Can {:?} fly? {}",
                maybe_flyer,
                maybe_flyer.can_fly()
            );
        }
        if let Some(legged) = context.get_controller_service_api(&NUMBER_OF_LEGS)? {
            critical!(logger, "{:?} has {} legs", legged, legged.number_of_legs());
        }
        Ok(OnTriggerResult::Ok)
    }
}

impl ProcessorDefinition for ZooProcessorRs {
    const DESCRIPTION: &'static str = "RUST TEST PROCESSOR: ZooProcessorRs";
    const INPUT_REQUIREMENT: ProcessorInputRequirement = ProcessorInputRequirement::Forbidden;
    const SUPPORTS_DYNAMIC_PROPERTIES: bool = false;
    const SUPPORTS_DYNAMIC_RELATIONSHIPS: bool = false;
    const OUTPUT_ATTRIBUTES: &'static [OutputAttribute] = &[];
    const RELATIONSHIPS: &'static [Relationship] = &[];
    const PROPERTIES: &'static [PropertyDefinition] =
        property_definitions![CAN_FLY_SERVICE, NUMBER_OF_LEGS];
}
